cpp/include/cudf/reduction.hpp

```diff
@@ -69,6 +72,46 @@ std::unique_ptr<scalar> reduce(
   data_type output_dtype,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief  Compute reduction of each segment in the input column
+ *
+ * This function does not detect overflows in reductions.
+ * Using a higher precision `data_type` may prevent overflow.
+ * The null values are skipped for the operation.
+ * If the segment is empty, the row corresponding to the result of the
+ * segment is null.
+ *
+ * @throw cudf::logic_error if reduction is called for non-arithmetic output
+ * type and operator other than `min` and `max`.
+ * @throw cudf::logic_error if input column data type is not convertible to
+ * output data type.
+ * @throw cudf::logic_error if `min` or `max` reduction is called and the
+ * output type does not match the input column data type.
+ * @throw cudf::logic_error if `any` or `all` reduction is called and the
+ * output type is not bool8.
+ *
+ * If the input column has arithmetic type, output_dtype can be any arithmetic
+ * type. If the input column has non-arithmetic type, e.g. timestamp or string,
+ * the same output type must be specified.
+ *
+ * @param col Input column view
+ * @param offsets Indices to segment boundaries
```

Comment:
Some more details here might be nice. Is this the same as the offsets in a list column? i.e. the first entry is always 0.

From the code it looks like the number of entries returned is determined by the number of offsets. Is there any issue if the offsets do not include everything in `col`? I assume not, but just want to be sure that the contract is clear.

What happens if the offsets are empty?  Will the resulting value be null? is it aggregation dependent?

Reply:
> Is this the same as the offsets in a list column? i.e. the first entry is always 0.

This API expect input has at least 1 segment (1 row in the list column), the rest is the same with the usual offsets column.

> Is there any issue if the offsets do not include everything in col?

It's ok. Ultimately we use a pair of neighboring items in the offset column to specify begin and end range of elements to reduce for. As long as the indices doesn't go beyond the data boundary it should be well defined.

And this is a test case that's missing at the moment.

> What happens if the offsets are empty?

Do you mean if `offsets` column is empty? It's currently rejected as invalid input.

Will update docstrings accordingly.

Reply:
> This API expect input has at least 1 segment (1 row in the list column), the rest is the same with the usual offsets column.

Revise - when you have 0 rows and 1 item in the offset vector (0), this is also valid input.